core/src/event.rs

```diff
@@ -61,7 +80,14 @@ impl Consumer {
         if let EventSubscriberMessage::EventReceived = message.into_v1() {
             Ok(())
         } else {
-            Err(eyre!("Expected `EventReceived`."))
+            Err(Error::ExpectedEventReceived)
         }
     }
```

Comment:
I assume that when socket is closed on the client side an error return value will be generated in this function. Couldn't you catch that error here and close the stream?

Reply:
Maybe I don't clearly understand what you are meaning, but I think it's not this place.

Do you suggest to remove `tokio::select!` call from `handle_subscription()` and make processing here? I think that would be incorrect.

In this function `Consumer` just waits client confirmation about successfully received event.
Does client need to wait for event to just say to server that it no longer needs new events and want to close conneciton? I think no. Client should be able to close connection at any time. So wee need to listen for new events from `events` and for new messages from client `stream` at the same time. That's why I used `tokio::select!`.

If you are talking about incorrect client behaviour when it does not send confirmation about received event and just closes its socket... When it should be processed normally:

1. `self.stream.recv().await?` will fail
2. `consumer.consume(evet).await?` also will fail
3. `handle_subscription()` will be ended with report about closed connection
4. API router will log it

Reply:
What Marin means here, is that while the result of the evaluation has an `Err` variant, it's not necessarily something we should propagate up the call chain. Closing a websocket connection is a normal part of the operation of Iroha, and it should probably be processed in-place.

Reply:
Then how should `handle_subscription()` stop working if it doesn't know about unexpected socket closing?

Reply:
Well, a good hint is that sometimes the closing of the socket is expected, and that is not an error. If the socket closure happens abruptly, it's a different case.

Reply:
Okay, now I understand what you are talking about.

I didn't change part with client response receiving:

```rust
let message: VersionedEventSubscriberMessage = self.stream.recv().await?;
```

So, do you think I should process all parts like that for unexpected socket closing? That also can happen in `Consumer::new()`. If yes, then how can I tell calling function, that `stream` is no longer available? Returning `Result<bool>`? Kinda wierd for that purpose.
WebSocket *closing handshake* is not just socket closing, so I think it's an erroneous client behaviour. It's like unexpected connection failure

Reply:
I've changed code to handle `Close` message in one place. So I'll mark this conversation as resolved and you can leave some comments for new code version